### ISSUE TITLE
[Request] Own CPU usage statistics

### ISSUE THREAD

**Issue by GuilhermeGS2** [hrydgard/ppsspp#10406]
I'd like something that shows the total CPU usage of usable CPU by the emulator like the EE percentage in PCSX2. Is it possible?

**Comment by hrydgard** [comment]
We have this implemented as a compile time option (enabled USE_PROFILER in ext/native/profiler/profiler.h) but it also shows an overlay with lots of details and since it's a compile time option you need to make a custom build for it.

Might one day make it or part of it a togglable feature...

**Comment by GuilhermeGS2** [comment]
Fine then. Thaks for the reply!

**Comment by zminhquanz** [comment]
That's interesting thing

**Comment by orbea** [comment]
Another way to do this for anyone using the gallium backend with mesa (amd and nouveau users) can use the GALLIUM_HUD. For example.
```
GALLIUM_HUD=cpu glxgears
```

**Comment by LunaMoo** [comment]
Experimented a bit with providing simpler and less covering info from Frame Profiler:
![npjh50414_00005](https://user-images.githubusercontent.com/5485237/34082306-b3c69414-e35b-11e7-9881-0d9a69ff570f.jpg)
 It works fine as a normal option and as long as it's not actually activated it doesn't really make a difference, however when activated, it tanks performance pretty heavily on it's own. Not sure if it's caused by the timing or just text refreshes on the screen, but it's pretty heavy.


 I also feel the situation of PCSX2 is way different, they have extremely heavy cpu thread that drag everything down and also have a hack which can put a lot of that load on a separate thread in which case some basic info can quickly tell whenever using such or other cpu speedhack actually helps or does the opposite.
 In PPSSPP most bottlenecks will be on the gpu or gpu driver side, so outside of few things like Final Fantasy IV splines(thundaga) or Juiced 2 beziers(roads) and similar effects, I don't feel like some cryptic detailed info is actually useful to end users;].

**Comment by GuilhermeGS2** [comment]
@LunaMoo ok then. Well, for most Android devices with Cortex A53 the main bottleneck is the CPU, I can even use 3xPSP resolution without change the performance in some games with low-end Mali T720MP2. Using some monitoring app I noticed a 100% CPU load in one core while others are idle. I really hope this future multithreading for OpengGL work well.

**Comment by LunaMoo** [comment]
CPU bottleneck on PPSSPP will most of the time be caused by your GPU drivers(still runs on cpu) and I don't think we can measure that. Pcsx2 doesn't either, in some ps2 games all shown threads will show small %, but game will run very poorly, yet if you check GPU usage it will not be high either - that's typically an indication of gpu driver bottleneck.

 Instead of looking at some % of cpu threads that most of the time will be very light in PPSSPP, you can turn on debug statistics and look at different numbers like draw calls or vertices as most gpu's will not handle well very high numbers of those which many popular (and demanding) games will abuse.

**Comment by GuilhermeGS2** [comment]
The debug option is available on Android too. I still have slow down in games with less than 400 draw calls, God of War Ghost of Sparta reaches 4500 draw calls 2xPSP res and 100% GPU usage (menu). I know Mali drivers are buggy, but my revision r7p0 isn't as buggy as newer ones, was the last good in performance.

**Comment by unknownbrackets** [comment]
CPU usage is kinda hard to measure accurately, and measuring it also slows it down (almost like a Heisenberg principal.)  The more accurate we make it (especially to exclude GPU driver CPU usage), the more it will slow things down (as LunaMoo noticed.)

If you want to identify how much faster the game would run if the jit did even more optimization, you could crank up frameskip, then look at system CPU monitors.  Anyway, frameskip never skips any CPU emulation, so if frameskip doesn't help, your problem is CPU.

Note: your problem can also be memory bandwidth.  Cranking up frameskip will also remove the GPU's load on unified memory bandwidth.  Usually this will be measured as CPU time.

If you want to identify how much faster the game would run on a device with a faster CPU, then the GPU driver would get faster too.  So, existing CPU usage monitors are probably the way to go.  But note that mobile devices will almost never have the same speed of GPU attached to a faster CPU anyway...

-[Unknown]

**Comment by GuilhermeGS2** [comment]
@unknownbrackets thanks for the clarification.

**Comment by Michal1324** [comment]
@unknownbrackets Yes but even with Android 7.0 you can't have active monitoring with CPU-Z when using split screen because when window isn't active it doesn't update itself but it should

Also switching active window will crash PPSSPP

**Comment by leopard20** [comment]
You should use monitors that use "drawing over other apps". I personally use Cool Tool.

**Comment by Gamemulatorer** [comment]
This might be related to https://github.com/hrydgard/ppsspp/issues/13220
